Android: add native platform support

buttaface opened the pull request:
This pull was used to build SPM natively on Android in [the Termux app](https://termux.com/), similar to apple/swift#27167. The first commit consists of compile-time changes that were needed to build SPM itself natively on an Android host, whereas the second commit adds some support for using SPM to build for Android. None of the second commit was needed to run the SPM tests on Android, so any of it can be removed if it interferes with any other plans for Android compilation, though there's currently only two mentions of Android in the SPM source.

This pull was tested against the 10/24 source snapshot and out of 623 test cases run, it reports 72 failing. Most are related to looking for PackageDescription4 and other dependencies of SPM itself, [which fails here because it can't find them relative to the test case's temporary path](https://github.com/apple/swift-package-manager/blob/e266c21086e36e16fc05e9af85e223c7afbe6c01/Sources/PackageLoading/PackageBuilder.swift#L482). A couple `print`s show that that code path isn't hit on linux, but I'm not familiar enough with the SPM source yet to fix that. In the meantime, I thought I'd get the rest of these changes in.

I also had to disable interpreting the SPM manifest to generate the JSON version when bootstrapping, similar to #2363, instead extracting the same command and generating the JSON by compiling the source. Finally, I had to tweak the bootstrap script:
```
diff --git a/Utilities/bootstrap b/Utilities/bootstrap
index 765b2488..afd9b09c 100755
--- a/Utilities/bootstrap
+++ b/Utilities/bootstrap
@@ -265,6 +265,7 @@ class Target(object):
         if args.foundation_path:
             import_paths.append(args.foundation_path)
             other_args.extend(["-Xcc", "-F" + args.foundation_path])
+            other_args.extend(["-Xcc", "-D__ANDROID_API__=23", "-Xcc", "-U_GNU_SOURCE"])
             import_paths.append(os.path.join(args.foundation_path, "swift"))
         if args.libdispatch_build_dir:
             import_paths.append(os.path.join(args.libdispatch_build_dir, 'lib'))
@@ -583,6 +584,7 @@ class llbuild(object):
                 if platform.system() == 'Linux':
                     link_command.extend(
                         ["-Xlinker", "-rpath=$ORIGIN/../lib/swift/linux"])
+                    link_command.extend(["-Xlinker", "-landroid-spawn"])
                 if self.args.foundation_path:
                     link_command.extend([
                         "-L", self.args.foundation_path,
@@ -1063,7 +1065,15 @@ def main():
     # Compute the build paths.
     if platform.system() == 'Darwin':
         build_target = "x86_64-apple-macosx"
-    elif platform.system() == 'Linux':              
+    elif platform.system() == 'Linux':
+        if 'ANDROID_DATA' in os.environ:
+            if platform.machine().startswith("armv7"):
+                build_target = 'armv7-unknown-linux-androideabi'
+            elif platform.machine() == 'aarch64':
+                build_target = 'aarch64-unknown-linux-android'
+            else:
+                raise SystemExit("ERROR: unsupported Android platform:",platform.machine())
+
         if platform.machine() == 'x86_64':
             build_target = "x86_64-unknown-linux-gnu"
         elif platform.machine() == "i686":
@@ -1075,7 +1085,7 @@ def main():
         elif platform.machine().startswith("armv7"):
             build_target = 'armv7-unknown-linux-gnueabihf'
         elif platform.machine() == 'aarch64':
-            build_target = 'aarch64-unknown-linux'
+            build_target = 'aarch64-unknown-linux-android'
         else:
             raise SystemExit("ERROR: unsupported machine:",platform.machine())
     elif platform.system() == 'Windows':
@@ -1269,6 +1279,7 @@ def main():
         else:
             embed_rpath = "@executable_path/../lib/swift/macosx"
     build_flags.extend(["-Xlinker", "-rpath", "-Xlinker", embed_rpath])
+    build_flags.extend(["-Xlinker", "-landroid-spawn"])
     if args.verbose:
         build_flags.append("-v")
 
@@ -1282,6 +1293,8 @@ def main():
         # Pass -F flag for CoreFoundation.
         build_flags.extend(["-Xswiftc", "-Xcc", "-Xswiftc", "-F{}".format(faketoolchain_includedir)])
 
+        build_flags.extend(["-Xswiftc", "-Xcc", "-Xswiftc", "-D__ANDROID_API__=23"])
+        build_flags.extend(["-Xswiftc", "-Xcc", "-Xswiftc", "-U_GNU_SOURCE"])
         # Pass -I for underlying Dispatch module.
         build_flags.extend(["-Xswiftc", "-I{}".format(os.path.join(faketoolchain_includedir, "dispatch"))])
 ```
The two preprocessor definitions are needed because [Bionic's `strerror_r` used to return a different type before API 23](https://android.googlesource.com/platform/bionic/+/master/libc/include/string.h#96); linking against android-spawn because Termux backported and shipped `posix_spawn` in a separate library; and I'm unsure why that last `build_target` change was needed, given the prior change for Android (without it, two directories are created with both names and the build errors out). I plan to try cross-compiling SPM itself for Android, so I'll submit a pull for all those build changes later.

Two other minor issues while I'm here:
- The build output warns against [calling `-fmodules-cache-path` on clang from the bootstrap script](https://github.com/apple/swift-package-manager/blob/3dcd3a314a0bbe6bafad315126d2d810bd8da1ff/Utilities/bootstrap#L300), on both Android and [linux](https://forums.swift.org/t/swift-build-failure-on-aur-package/30328):
```
clang-8: warning: argument unused during compilation: '-fmodules-cache-path=/home/pacman/aur/swift-language/src/build/Ninja-ReleaseAssert/swiftpm-linux-x86_64/.bootstrap/ModuleCache' [-Wunused-command-line-argument]
```
I'm guessing it's expected that Swift's forked clang is used, but unless it's installed in the system itself, the bootstrap script isn't configured that way. I'm unsure if this was intended or if it matters.
- I don't think the bootstrap script compiles in parallel, would be worthwhile to add that to speed up full rebuilds. I saw nice speed gains when Saleem added that for Foundation, apple/swift-corelibs-foundation#2129.

aciidb0mb3r commented on Sources/PackageDescription4/SupportedPlatforms.swift:
```diff
@@ -37,6 +37,9 @@ public struct Platform: Encodable {
     public static let windows: Platform = Platform(name: "windows")
+
+    /// The Android platform
+    public static let android: Platform = Platform(name: "android")
```
Please annotate this with `@available(_PackageDescription, introduced: 5.2)`

buttaface replied:
Done.

buttaface commented on Sources/Build/Triple.swift:
```diff
@@ -129,6 +133,7 @@ public struct Triple: Encodable {
     public static let arm64Linux = try! Triple("aarch64-unknown-linux")
     public static let armLinux = try! Triple("armv7-unknown-linux-gnueabihf")
     public static let android = try! Triple("armv7-unknown-linux-androideabi")
```
This was added years back and never used anywhere, I can change it to `armAndroid` to match the other variables' naming if wanted.

aciidb0mb3r replied:
Sure

buttaface commented on Sources/PackageModel/Platform.swift:
```diff
@@ -55,6 +55,7 @@ public struct Platform: Equatable, Hashable {
     public static let tvOS: Platform = Platform(name: "tvos", oldestSupportedVersion: "9.0")
     public static let watchOS: Platform = Platform(name: "watchos", oldestSupportedVersion: "2.0")
     public static let linux: Platform = Platform(name: "linux", oldestSupportedVersion: .unknown)
+    public static let android: Platform = Platform(name: "android", oldestSupportedVersion: "7.0")
```
[The Termux package repo supports back to 7.0](https://github.com/termux/termux-packages/blob/master/README.md#project-structure) and three years back seemed a reasonable default, but I don't care what version is used here.

jakepetroules replied:
I'm a little concerned about committing to a versioning scheme here.

First, the backwards compatibility story works a bit differently on Android, where there are two versions to worry about at build time: `minSdkVersion` (roughly equivalent to the Apple "deployment target" concept), and `targetSdkVersion` (roughly equivalent to the Apple SDK version, but on Android you CAN change this, on Apple platforms it's fixed by a particular Xcode version). There's also maxSdkVersion, but this is deprecated and ignored nowadays. See https://developer.android.com/guide/topics/manifest/uses-sdk-element.html#target for more info.

Secondly, developers don't use the Android marketing versions like "7.0", "8.0", etc., for anything. They use [API levels](https://source.android.com/setup/start/build-numbers) instead, and no one has decided yet how Swift availability will work on Android (whether it would use API levels like native Android tools, or whether it would attempt to map marketing versions to API levels).

Let's leave this at "unknown" until we have a better idea of how this should work in the Swift world.

buttaface commented on Sources/TSCUtility/FSWatch.swift:
```diff
@@ -428,55 +428,74 @@ public final class Inotify {
     let intOffset = Int(fd / 16)
     let bitOffset = Int(fd % 16)
-    let mask = 1 << bitOffset
+    let mask: UInt = 1 << bitOffset
```
I'd get errors that `mask` was set to `Int` otherwise, unsure why this varies for Android AArch64 but just specified the type.

aciidb0mb3r commented:
Thanks a lot for working on this! This looks good to me. I think @gmittert is working on adding support for compiling the manifest file for Windows so that should be helpful for Android as well.

aciidb0mb3r commented:
@swift-ci smoke test

buttaface commented on Sources/TSCUtility/FSWatch.swift:
```diff
@@ -428,55 +428,74 @@ public final class Inotify {
+    set.fds_bits = fd_bits
+  #else
+    set.__fds_bits = fd_bits
+  #endif
```
It'd be nice if I could assign some kind of compile-time symbol reference to each platform's variable name, so that this runtime copying isn't needed.

buttaface commented on Sources/TSCUtility/Platform.swift:
```diff
@@ -33,6 +34,10 @@ public enum Platform {
+            if localFileSystem.isFile(AbsolutePath("/system/bin/toolbox")) ||
+               localFileSystem.isFile(AbsolutePath("/system/bin/toybox")) {
+                return .android
+            }
```
This function doesn't seem to be called anywhere, can remove if not needed.

aciidb0mb3r replied:
yah let's remove it if it's dead code.

buttaface replied:
I was wrong, the variable it's used to initialize, `currentPlatform`, is used [in exactly one place](https://github.com/apple/swift-package-manager/blob/8b6eebd3d83fb0dd6e26ffba195ba5bcb143c356/Sources/PackageLoading/Target%2BPkgConfig.swift#L111). I suppose this might be useful for TSCUtility and llbuild someday too, [for which I've created a Termux package](https://github.com/termux/termux-packages/tree/master/packages/llbuild), so I'd like to keep this Android support.

jakepetroules replied:
Any reason you chose these particular files vs `/system/build.prop`?

buttaface replied:
No other reason that I'm familiar with toolbox/toybox, as they're sometimes invoked from the Termux command-line, and not build.prop, which is unreadable without root on my device. If there's a better file that signifies Android, glad to use that instead, just let me know.

jakepetroules replied:
OK, that's fine for now, we can always update the check later if we need to.

Thanks for your contribution!

buttaface commented on Sources/TSCUtility/dlopen.swift:
```diff
@@ -60,8 +60,10 @@ public struct DLOpenFlags: RawRepresentable, OptionSet {
   #else
     public static let first: DLOpenFlags = DLOpenFlags(rawValue: 0)
+  #if !os(Android)
     public static let deepBind: DLOpenFlags = DLOpenFlags(rawValue: RTLD_DEEPBIND)
```
No `RTLD_DEEPBIND` on Android, so #ifdef it out here and above.

buttaface commented:
Hmm, the failed smoke test indicates that the expected type of `mask` varies by platform, I'll investigate further tonight and fix it.

jakepetroules commented on Sources/Build/Triple.swift:
```diff
@@ -129,6 +133,7 @@ public struct Triple: Encodable {
     public static let arm64Linux = try! Triple("aarch64-unknown-linux")
     public static let armLinux = try! Triple("armv7-unknown-linux-gnueabihf")
     public static let android = try! Triple("armv7-unknown-linux-androideabi")
+    public static let arm64Android = try! Triple("aarch64-unknown-linux-android")
```
Add the x86 and x86_64 ABIs as well?

jakepetroules commented on Tests/PackageLoadingTests/PackageBuilderTests.swift:
```diff
@@ -1376,6 +1376,7 @@ class PackageBuilderTests: XCTestCase {
             "ios": "8.0",
             "tvos": "9.0",
             "watchos": "2.0",
+            "android": "7.0",
```
Remove

buttaface replied:
Had to set to 0.0 just like linux above for this test to pass.

jakepetroules commented on Tests/PackageLoadingTests/PackageBuilderTests.swift:
```diff
@@ -1411,6 +1412,7 @@ class PackageBuilderTests: XCTestCase {
             "linux": "0.0",
             "ios": "8.0",
             "watchos": "2.0",
+            "android": "7.0",
```
Remove

buttaface commented:
Alright, figured out the issue with the type of `mask`: it's that [`fd_mask` is a UInt in Bionic](https://android.googlesource.com/platform/bionic/+/master/libc/include/sys/select.h#44) but [Int in glibc](https://sourceware.org/git/?p=glibc.git;a=blob;f=misc/sys/select.h;h=4e47f3fe1334ed63140687e1f7d6b7a04423c331;hb=HEAD#l49), moved `mask` inside the #ifdef to account for that.

Should be ready to go now, addressed the other issues too. The sourcekit-lsp pull will need to be tested and merged alongside this.

aciidb0mb3r commented:
@swift-ci smoke test

aciidb0mb3r commented:
test

aciidb0mb3r commented:
test

aciidb0mb3r commented:
test

aciidb0mb3r commented:
test

aciidb0mb3r commented:
Please test with the following pull requests:
https://github.com/apple/sourcekit-lsp/pull/186

@swift-ci smoke test

jakepetroules commented on Sources/Build/Triple.swift:
```diff
@@ -128,7 +132,10 @@ public struct Triple: Encodable {
     public static let arm64Linux = try! Triple("aarch64-unknown-linux")
     public static let armLinux = try! Triple("armv7-unknown-linux-gnueabihf")
-    public static let android = try! Triple("armv7-unknown-linux-androideabi")
+    public static let armAndroid = try! Triple("armv7-unknown-linux-androideabi")
```
One last thing, the [target triple](https://developer.android.com/ndk/guides/other_build_systems) should be `armv7a` (sorry I missed that before)

buttaface replied:
Will do.

aciidb0mb3r commented:
Please test with the following pull requests:
https://github.com/apple/sourcekit-lsp/pull/186

@swift-ci smoke test

buttaface commented:
Thanks for the quick review and merge, got many of those tests working in the meantime with an adapted version of #2363, will submit that later.
